### PR TITLE
Add `prefer-dataset` rule

### LINKED ISSUE #166: Rule proposal: `prefer-dataset`
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/166)

Prefer the [`.dataset`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset) property on DOM elements over using `setAttribute`.

We would detect `setAttribute` and that the string starts with `data-`.

## Fail

```js
element.setAttribute('data-unicorn', '🦄');
```

## Pass

```js
element.dataset.unicorn = '🦄';
```

> There is a $60.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/166).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by amedora**
Fix #166

### CODE REVIEW

File: rules/prefer-dataset.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+	return `${objectName}.dataset.${propertyName} = ${value}`;
+};
+
+const create = context => {
+	return {
+		CallExpression(node) {
+			const {callee} = node;
+
+			if (callee.type === 'MemberExpression' && getMethodName(callee) === 'setAttribute') {
```
Comment (futpib) [inline rules/prefer-dataset.js]: Code like `element[setAttribute]('data-unicorn', '🦄')` will "break" this check

Comment (amedora) [reply rules/prefer-dataset.js]: oh, I got you.
I have no idea which one is better...
- just ignore bracket notation
- fix it to element[dataset].unicorn = '🦄';

Comment (sindresorhus) [reply rules/prefer-dataset.js]: You have to ignore bracket notation. It’s very hard to impossible to reliably check this.

Comment (sindresorhus) [reply rules/prefer-dataset.js]: You should add some tests for this case too.

**Comment by sindresorhus** [comment]
Hey, thanks for the pull request 🙌 I have gotten a huge response on the IssueHunt bounty issues and I now have a very long backlog of pull requests to review, so it might take some time until I will be able to review this. However, you can help move this along by helping out and reviewing the [other submissions](https://github.com/sindresorhus/eslint-plugin-unicorn/pulls). If everyone helps everyone we can manage this 👌.

Some things to look for when reviewing:

- Ensure the rule addition adheres to the [instructions](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/new-rule.md).
- There are never too many tests. Try to find both passing and failing cases, especially obscure cases.
- Check the docs for typos and whether something could be more clear.
- Look at the rule code and see if anything could be simplified or done better. If something is unclear in the code, ask the submitter to explain.

**Comment by sindresorhus** [comment]
Since almost nobody actually volunteered to review other PRs, I've decided to make it a requirement. You have to review at least two other PRs to get your own PR merged. Thank you for understanding.

https://github.com/sindresorhus/eslint-plugin-unicorn/pull/226#issuecomment-459012736

File: test/prefer-dataset.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+		'element.dataset[\'unicorn\'] = \'🦄\';',
+		'element[setAttribute](\'data-unicorn\', \'🦄\');',
+		'element.setAttribute(\'foo\', \'bar\');'
+	],
+	invalid: [
+		{
+			code: 'element.setAttribute(\'data-unicorn\', \'🦄\');',
+			errors: [{message: 'Prefer `dataset` over `setAttribute`'}],
+			output: 'element.dataset.unicorn = \'🦄\';'
+		}
```
Comment (MrHen) [inline test/prefer-dataset.js]: A few more test cases worth checking:

* `element.setAttribute(\'data-foo-bar\', \'zaz\')`
* `element.setAttribute(\'data-🦄\', \'🦄\')`
* `element.setAttribute(foo, bar)`
* `element.setAttribute(\'data-foo2\', \'bar\')`
* `element.setAttribute(\'data-foo:bar\', \'zaz\')`
* `element.setAttribute(\'data-foo\', /* comment */ \'bar\')`
* ```
element.setAttribute(
  \'data-foo\', // comment
  \'bar\', // comment
)
```

Note that the fix conversions should follow the description in the [documentation](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset):

> The name of a custom data attribute in JavaScript is the name of the same HTML attribute but in camelCase and with no dashes, dots, etc.

I think this means `element.setAttribute('data-date-of-birth', '1960-10-03')` should be fixed to `element.dataset.dateOfBirth = '1960-10-03'`.

Comment (amedora) [reply test/prefer-dataset.js]: Thank you for your feedback :smiley:

File: test/prefer-dataset.js [hunk 3]
```diff
@@ -8,18 +9,67 @@ const ruleTester = avaRuleTester(test, {
...
+		},
+		{
+			code: 'element.setAttribute(\'data-foo-bar\', \'zaz\');',
+			errors,
+			output: 'element.dataset.fooBar = \'zaz\';'
+		},
+		{
+			code: 'element.setAttribute(\'data-foo\', /* comment */ \'bar\');',
+			errors,
+			output: 'element.dataset.foo = \'bar\';'
```
Comment (MrHen) [inline test/prefer-dataset.js]: I don't think we want to delete comments during fixing but it probably doesn't come up that often. @sindresorhus, what's your take on this?

1. Removing comments is okay
2. Comments should be preserved (may be non-trivial)
3. Do not fix if comments are detected

[I've done (3) in the past](https://github.com/sindresorhus/eslint-plugin-unicorn/pull/305/files#diff-c717db2a17e1fd6455af92ec118c404eR35) but I think a bunch of different rules current do (1).

Comment (sindresorhus) [reply test/prefer-dataset.js]: @MrHen I don't think we should do `1.`, but we need to come up with a guideline that we can apply to all rules. Maybe we could make an utility that just extract all comments to above a node we give it? The comments can require manual formatting and movement. We just shouldn't remove them. Alternatively, `3.` if `2.` is too complex, but I think we could make something generic that we could apply to many rules. Would you be able to open a new issue about it? We probably need to review all existing rules for this and update the contributing guidelines.

**Comment by sindresorhus** [comment]
Thanks, @amedora :)

**Comment by sindresorhus** [comment]
Note https://github.com/sindresorhus/eslint-plugin-unicorn/commit/850b8a1eb3af5edf9c02c440672a8c293231429c. I missed that during the review.

File: test/prefer-dataset.js (hunk 3 above)
Comment (MrHen) [reply test/prefer-dataset.js]: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/370

**Pull request merged**